src/main/java/com/mongodb/jdbc/oidc/OidcAuthFlow.java

```diff
@@ -0,0 +1,159 @@
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.jdbc.oidc;
+
+import com.nimbusds.oauth2.sdk.AuthorizationCode;
+import com.nimbusds.oauth2.sdk.AuthorizationCodeGrant;
+import com.nimbusds.oauth2.sdk.AuthorizationRequest;
+import com.nimbusds.oauth2.sdk.ResponseType;
+import com.nimbusds.oauth2.sdk.Scope;
+import com.nimbusds.oauth2.sdk.TokenRequest;
+import com.nimbusds.oauth2.sdk.TokenResponse;
+import com.nimbusds.oauth2.sdk.http.HTTPResponse;
+import com.nimbusds.oauth2.sdk.id.ClientID;
+import com.nimbusds.oauth2.sdk.id.Issuer;
+import com.nimbusds.oauth2.sdk.id.State;
+import com.nimbusds.oauth2.sdk.pkce.CodeChallengeMethod;
+import com.nimbusds.oauth2.sdk.pkce.CodeVerifier;
+import com.nimbusds.oauth2.sdk.token.Tokens;
+import com.nimbusds.openid.connect.sdk.OIDCTokenResponse;
+import com.nimbusds.openid.connect.sdk.OIDCTokenResponseParser;
+import com.nimbusds.openid.connect.sdk.op.OIDCProviderMetadata;
+import java.awt.*;
+import java.net.URI;
+import java.time.Duration;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class OidcAuthFlow {
+
+    private static final Logger logger = Logger.getLogger(OidcAuthFlow.class.getName());
+
+    public OidcCallbackResult doAuthCodeFlow(OidcCallbackContext callbackContext) {
+        IdpInfo idpServerInfo = callbackContext.getIdpInfo();
+        if (idpServerInfo == null) {
+            logger.severe("IdpServerInfo is null");
+            return null;
+        }
+
+        String clientID = idpServerInfo.getClientId();
+        if (clientID == null || clientID.isEmpty()) {
+            logger.severe("Client ID is null or empty");
+            return null;
+        }
+
+        String issuerURI = idpServerInfo.getIssuer();
+        if (!issuerURI.startsWith("https")) {
+            logger.severe("Issuer URI must be HTTPS");
+            return null;
+        }
+
+        List<String> scopesList = idpServerInfo.getRequestScopes();
+        Scope scopes = new Scope();
+        if (scopesList != null) {
+            for (String scope : scopesList) {
+                scopes.add(new Scope.Value(scope));
+            }
+        }
+        // mongodb is not configured to ask for offline_access by default. We prefer always getting a
+        // refresh token when the server allows it.
+        scopes.add(new Scope.Value("offline_access"));
+
+        RFC8252HttpServer server = new RFC8252HttpServer();
+        try {
+            // Resolve OIDC provider metadata using the issuer URI and
+            // extract authorization and token endpoint URIs.
+            OIDCProviderMetadata providerMetadata =
```

Comment:
We should make sure that scopes we are requesting are in the supported scopes: https://www.javadoc.io/doc/com.nimbusds/oauth2-oidc-sdk/5.0/com/nimbusds/openid/connect/sdk/op/OIDCProviderMetadata.html#getScopes()

For instance, if the IdP does not support offline_access (refresh) we don't want to just fail, we would prefer just not being able to refresh.

Reply:
I added a change to check to use getScopes() on the metadata and add the requested scope if it's supported.